asserts/snapasserts/snapasserts_test.go

```diff
@@ -789,3 +791,322 @@ provenance: prov`, nil)
 	_, err = snapasserts.DeriveSideInfoFromDigestAndSize(withProv, digest, size, nil, s.localDB)
 	c.Check(err, ErrorMatches, `safely handling snaps with different provenance but same hash not yet supported`)
 }
+
+func (s *snapassertsSuite) TestCrossCheckResource(c *C) {
+	digest := makeDigest(12)
+	componentRev := snap.R(24)
+	snapRev := snap.R(12)
+	const size = uint64(1024)
+	const resourceName = "test-component"
+	const snapID = "snap-id-1"
+
+	revHeaders := map[string]interface{}{
+		"snap-id":           snapID,
+		"resource-name":     resourceName,
+		"resource-sha3-384": digest,
+		"resource-revision": componentRev.String(),
+		"resource-size":     strconv.Itoa(int(size)),
+		"developer-id":      s.dev1Acct.AccountID(),
+		"timestamp":         time.Now().Format(time.RFC3339),
+	}
+
+	resourceRev, err := s.storeSigning.Sign(asserts.SnapResourceRevisionType, revHeaders, nil, "")
+	c.Assert(err, IsNil)
+	err = s.localDB.Add(resourceRev)
+	c.Assert(err, IsNil)
+
+	pairHeaders := map[string]interface{}{
+		"snap-id":           snapID,
+		"resource-name":     resourceName,
+		"resource-revision": componentRev.String(),
+		"snap-revision":     snapRev.String(),
+		"developer-id":      s.dev1Acct.AccountID(),
+		"timestamp":         time.Now().Format(time.RFC3339),
+	}
+
+	resourcePair, err := s.storeSigning.Sign(asserts.SnapResourcePairType, pairHeaders, nil, "")
+	c.Assert(err, IsNil)
+	err = s.localDB.Add(resourcePair)
+	c.Assert(err, IsNil)
+
+	si := &snap.SideInfo{
+		SnapID:   "snap-id-1",
+		Revision: snap.R(12),
+	}
+
+	csi := &snap.ComponentSideInfo{
+		Component: naming.NewComponentRef("snap", "test-component"),
+		Revision:  snap.R(24),
+	}
+
+	// everything cross checks, with the regular snap name
+	err = snapasserts.CrossCheckResource(resourceName, digest, "", size, csi, si, nil, s.localDB)
+	c.Assert(err, IsNil)
+}
+
+func (s *snapassertsSuite) TestCrossCheckResourceMissingRevisionAssertion(c *C) {
+	s.testCrossCheckResourceMissingRevisionAssertion(c, "")
+}
+
+func (s *snapassertsSuite) TestCrossCheckResourceProvenanceMissingRevisionAssertion(c *C) {
+	s.testCrossCheckResourceMissingRevisionAssertion(c, "prov")
+}
+
+func (s *snapassertsSuite) testCrossCheckResourceMissingRevisionAssertion(c *C, provenance string) {
+	digest := makeDigest(12)
+	const size = uint64(1024)
+	const resourceName = "test-component"
+	const snapID = "snap-id-1"
```

Comment:
This isn't being used below. Also, as a minor preference, not having these externalised unless it's necessary makes it easier to follow the test IMO